File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.js
// input
const fs = require("fs");
const stdin = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const answer = [];

const makePIV = (n) => {
  let PIV = 1;
  while (PIV <= n) PIV *= 2;
  return PIV;
};

class IndexTree {
  constructor(PIV) {
    this.PIV = PIV;
    this.tree = Array.from({ length: PIV * 2 }, () => BigInt(0));
  }

  update(i, v) {
    let index = i + this.PIV;
    this.tree[index] = v;

    // 부모로 타고 올라가면서 값을 갱신한다.
    index = parseInt(index / 2);

    // 루트(1번 원소)까지
    while (index >= 1) {
      this.tree[index] = this.tree[2 * index] + this.tree[2 * index + 1];
      index = parseInt(index / 2);
    }
  }

  init(array) {
    for (let i = 0; i < array.length; i++) {
      this.update(i + 1, array[i]);
    }
  }

  // l에서 r까지의 구간 합
  query(l, r) {
    let sum = BigInt(0);
    let [left, right] = [l + this.PIV, r + this.PIV];

    while (left <= right) {
      if (left % 2 == 1) sum += this.tree[left++];
      if (right % 2 == 0) sum += this.tree[right--];

      [left, right] = [parseInt(left / 2), parseInt(right / 2)];
    }

    answer.push(BigInt(sum));
  }
}

const [n, m, k] = stdin[0].split(" ").map((num) => parseInt(num));
const numbers = stdin.slice(1, n + 1).map((num) => BigInt(num));
const ops = stdin
  .slice(n + 1)
  .map((row) => row.split(" ").map((num) => parseInt(num)));

const PIV = makePIV(n);
const tree = new IndexTree(PIV);
tree.init(numbers);

ops.forEach((op) => {
  op[0] === 1 ? tree.update(op[1], BigInt(op[2])) : tree.query(op[1], op[2]);
});

console.log(answer.join("\n"));

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.cc
#include <bits/stdc++.h>
using namespace std;

void update(int n, long long v, vector<long long> &tree, int PIV) {
  n += PIV;
  tree[n] = v;
  n /= 2;
  while (n > 0) {
    // 조상 = 왼쪽 자식 + 오른쪽 자식
    tree[n] = tree[n * 2] + tree[n * 2 + 1];
    n /= 2;  // 윗 조상으로 옮김
  }
}

long long query(int l, int r, vector<long long> &tree, int PIV) {
  long long ret = 0;
  l += PIV, r += PIV;
  while (l <= r) {
    if (l % 2 == 1) ret += tree[l++];  // l이 홀수 일 때
    if (r % 2 == 0) ret += tree[r--];  // r이 짝수 일 떄
    l /= 2, r /= 2;
  }
  return ret;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n, m, k;
  cin >> n >> m >> k;

  // PIV 는 n 을 넘는 2의 배수이다.
  int PIV = 1;
  while (PIV <= n) PIV *= 2;

  vector<long long> tree(PIV * 2);
  fill(tree.begin(), tree.end(), 0);

  // init
  for (int i = 1; i <= n; i++) {
    long long num;
    cin >> num;
    update(i, num, tree, PIV);
  }

  int a, b;
  long long c;

  for (int i = 0; i < m + k; i++) {
    cin >> a >> b >> c;

    if (a == 1) {
      update(b, c, tree, PIV);
    } else {
      cout << query(b, (int)c, tree, PIV) << '\n';
    }
  }

  return 0;
}
